### PR TITLE
Add Zip file functionality. Fixes #11413

### LINKED ISSUE #11413: Why isn't zip compression included for read_csv?
Is there a reason why the compression options for read_csv and other IO functions don't include zip?

### PR DESCRIPTION
**Pull request by lababidi**
closes #11413

This PR leverages Python's ZipFile functionality to automatically unzip files read into DataFrames using read_csv().

### CODE REVIEW

File: pandas/io/tests/test_parsers.py [hunk 1]
```diff
@@ -3809,13 +3809,26 @@ def test_decompression(self):
...
+            tmp = zipfile.ZipFile(path, mode='w')
+            tmp.writestr(file_name, data)
+            tmp.close()
+
+            result = self.read_csv(path, compression='zip')
+            tm.assert_frame_equal(result, expected)
+
+            result = self.read_csv(open(path, 'rb'), compression='zip')
+            tm.assert_frame_equal(result, expected)
+
```
Comment (jreback) [inline pandas/io/tests/test_parsers.py]: need a test for multiple files in a zip (e.g. assert that the ``ValueError`` is raised)

**Comment by jreback** [comment]
as an aside, would be interested in solving issues brought up in https://github.com/pydata/pandas/issues/11666 and partially addressed in #11677 (not merged)

**Comment by lababidi** [comment]
@jreback I would be happy to continue this to the other `read_*` functions. I like the idea of refactoring out the compression determination and the decompression steps, so that it can be used in all the `read_*` functions. I won't use the PR #11677 because it does not include tests or zip functionality. It also only focuses on `pickles`.

**Comment by lababidi** [comment]
@jreback Could you merge this in and I'll take care of the other requests?

File: pandas/io/parsers.py [hunk 2]
```diff
@@ -61,9 +61,9 @@ class ParserWarning(Warning):
 dtype : Type name or dict of column -> type, default None
     Data type for data or columns. E.g. {'a': np.float64, 'b': np.int32}
     (Unsupported with engine='python')
-compression : {'gzip', 'bz2', 'infer', None}, default 'infer'
+compression : {'gzip', 'bz2', 'zip', 'infer', None}, default 'infer'
```
Comment (jreback) [inline pandas/io/parsers.py]: can you add an explanation about .zip is only a single file

File: pandas/parser.pyx [hunk 3]
```diff
@@ -563,6 +563,18 @@ cdef class TextReader:
...
+                import zipfile
+                zip_file = zipfile.ZipFile(source)
+                zip_names = zip_file.namelist()
+
+                if len(zip_names) == 1:
+                    file_name = zip_names.pop()
+                    source = zip_file.open(file_name)
+
+                elif len(zip_names)>1:
+                    raise ValueError('Multiple files found in compressed '
```
Comment (jreback) [inline pandas/parser.pyx]: maybe just do ``else`` here (e.g. you can have 0 files in an archive?)

File: pandas/io/tests/test_parsers.py [hunk 4]
```diff
@@ -3809,13 +3809,37 @@ def test_decompression(self):
...
+
+        with tm.ensure_clean() as path:
+            file_names = ['test_file', 'second_file']
+            tmp = zipfile.ZipFile(path, mode='w')
```
Comment (jreback) [inline pandas/io/tests/test_parsers.py]: add in another test using ``tmp`` before adding other files (should fail as well)

File: pandas/io/tests/test_parsers.py [hunk 5]
```diff
@@ -3809,13 +3809,37 @@ def test_decompression(self):
...
             import bz2
+            import zipfile
         except ImportError:
-            raise nose.SkipTest('need gzip and bz2 to run')
+            raise nose.SkipTest('need zipfile, gzip and bz2 to run')
 
         data = open(self.csv1, 'rb').read()
         expected = self.read_csv(self.csv1)
 
         with tm.ensure_clean() as path:
```
Comment (jreback) [inline pandas/io/tests/test_parsers.py]: put the zipfile tests in a separate test.

Comment (jreback) [inline pandas/io/tests/test_parsers.py]: pls split out the gzip/bz2 into separate tests as well, thanks

Comment (jreback) [inline pandas/io/tests/test_parsers.py]: best would be to make a separate testing class for ``TestCompression``

**Comment by lababidi** [comment]
@jreback could you please merge this before you make anymore changes to the tests.

**Comment by jreback** [comment]
you need to rebase then I'll review

**Comment by lababidi** [comment]
@jreback Rebased. Thanks.

**Comment by jreback** [comment]
you need to force push. you should only have your commits. see [here](http://pandas.pydata.org/pandas-docs/stable/contributing.html#contributing-your-changes-to-pandas)

**Comment by lababidi** [comment]
@jreback I'm redoing this Pull Request. I will close this one and open a new PR.

**Comment by jreback** [comment]
ok, in the future, just push to the same one

**Comment by itzybitzy** [comment]
Can the same functionality be added to read_fwf() and other read_* methods?

**Comment by jreback** [comment]
#11666 is related
#12688 is where this could be done, e.g. the code is a spread out a bit. welcome for you to take a crack at it @itzybitzy

**Comment by jreback** [comment]
IOW, the compression interfaces need to be pulled out a bit from the parser code

**Comment by itzybitzy** [comment]
Sorry @jreback, I just don't have the skills to do that (yet). Is there something good for a beginner to work on? Documentation maybe? (Not even sure where the right place to ask this is.)

**Comment by jreback** [comment]
http://pandas.pydata.org/pandas-docs/stable/contributing.html

selects label of difficulty novice and you will see lots of issues

**Pull request closed without merging**